### PR TITLE
#1965 add cash register page

### LINKED ISSUE #1965: Add cash register page
## Is your feature request related to a problem? Please describe.

Add cash register page.

## Describe the solution you'd like

Add basic cash register page and navigation logic.

## Implementation

- Add new page `CashRegisterPage`
- Add page `cashRegisterInitialiser`
- Add cash register columns to `GetPageInfoColumns`.

## Describe alternatives you've considered

N/A.

## Additional context

See epic #1917.

### PR DESCRIPTION
**Pull request by wlthomson**
Fixes #1965.

Basic working prototype. Still needs some tidy up.

Next steps:

- Add functionality for adding new cash transactions.
- Add functionality for syncing out cash transactions.

## Change summary

See below.

## Testing

More comprehensive tests to come once main functionality is completed.

Steps to reproduce or otherwise test the changes of this PR:

- [ ] Cash register page displays payments correctly.
- [ ] Cash register page displays receipts correctly.

### Related areas to think about

N/A.

### CODE REVIEW

File: src/database/UIDatabase.js [hunk 1]
```diff
@@ -127,6 +127,8 @@ class UIDatabase {
         return results.filtered('type == $0', 'receipt');
       case 'Payment':
         return results.filtered('type == $0', 'payment');
+      case 'CashTransaction':
```
Comment (wlthomson) [inline src/database/UIDatabase.js]: A workaround the fact that realm doesn't allow concatenation of result objects. Seems a bit redundant to combine two existing types, but at the moment can't think of a better way of doing this.

File: src/pages/CashRegisterPage.js [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+};
+
+const mapDispatchToProps = (dispatch, ownProps) => ({
+  ...getPageDispatchers(dispatch, ownProps, 'Transaction', ROUTES.CASH_REGISTER),
+});
+
+const mapStateToProps = state => {
+  const { pages } = state;
+  const { cashRegister } = pages;
+  return cashRegister;
```
Comment (wlthomson) [inline src/pages/CashRegisterPage.js]: Note that a few props and potentially selectors remaining to be added.

File: src/pages/dataTableUtilities/getColumns.js [hunk 3]
```diff
@@ -33,6 +33,7 @@ const PAGE_COLUMN_WIDTHS = {
   prescriberSelect: [3, 3, 1],
   itemSelect: [1, 3, 1],
   patientHistory: [1, 3, 1, 3],
+  [ROUTES.CASH_REGISTER]: [1, 1, 1, 1, 1, 1, 1, 1],
```
Comment (wlthomson) [inline src/pages/dataTableUtilities/getColumns.js]: Just a useful default, to be updated once core functionality completed.

File: src/pages/dataTableUtilities/getColumns.js [hunk 4]
```diff
@@ -261,6 +272,14 @@ const COLUMNS = () => ({
     sortable: true,
     editable: false,
   },
+  [COLUMN_NAMES.PAYMENT_NAME]: {
+    type: COLUMN_TYPES.STRING,
+    key: COLUMN_KEYS.OTHER_PARTY_NAME,
+    title: tableStrings.name,
+    alignText: 'left',
+    sortable: false,
+    editable: false,
```
Comment (wlthomson) [inline src/pages/dataTableUtilities/getColumns.js]: Nothing sortable, editable at the moment. Another UI thing to be updated after core functionality done.

File: src/sync/incomingSyncUtils.js [hunk 5]
```diff
@@ -725,6 +725,8 @@ export const createOrUpdateRecord = (database, settings, recordType, record) =>
         enteredBy,
         otherParty,
         linkedRequisition,
+        option: database.getOrCreate('Options', record.optionID),
+        total: parseFloat(record.total),
```
Comment (wlthomson) [inline src/sync/incomingSyncUtils.js]: These fields are already added to the `Transaction` object, but looks like the incoming sync logic wasn't added.

Comment (joshxg) [reply src/sync/incomingSyncUtils.js]: yeah - just added when starting dispensing but hadn't done a lot of the sync stuff.

The `total` was added - at the moment we derive the total from all of the trans lines of the transaction quantity * price - that's gonna need to change soon as the payment trans lines won't have a quantity :(

File: src/pages/dataTableUtilities/getPageInfoColumns.js [hunk 6]
```diff
@@ -138,6 +141,36 @@ const PAGE_INFO_ROWS = (pageObject, dispatch, route) => ({
...
+  comment: {
+    title: 'Comment',
+    info: pageObject.comment,
+    onPress: null,
+  },
+  status: {
+    title: 'Status',
+    info: pageObject.status,
+    onPress: null,
+  },
```
[the remark this answers is not in the document]
Comment (wlthomson) [reply src/pages/dataTableUtilities/getPageInfoColumns.js]: Good point. I think I'll leave it for safety just until the page UI is completed, but will remove as part of tidying up props once main functionality is completed.

**Pull request merged**